### PR TITLE
Develop - Linux/SQL Code Review

### PR DESCRIPTION
**Pull request by myanneng**

### CODE REVIEW

File: linux_sql/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+## Introduction
+
+Cluster Monitor Agent is an internal tool that monitors the cluster resources. Irecords the hardware specifications of each node (server), and monitors its resource usage (e.g. CPU/memory) in realtime. The collected data is stored in a database found in a designated primary server node (Server no. 1). This tool allows for the Linux Cluster Administration (LCA) team to use the data to generate reports for future resource planning, such as adding or removing servers based on their usage and activity.
+
+## Architecture and Design
+![Generalized architecture for the Cluster Monitor Agent](assets/architecture.png)
```
Comment (Jarvis-deprecated-account) [inline linux_sql/README.md:8]: fix diagram

File: linux_sql/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+`ddl.sql` sets up the PostgreSQL database.
+
+The bash agent runs two scripts: `host_info.sh` and `host_usage.sh`.
+
+`host_info.sh` is a script that is run only once after the first initalization. It gathers the hardware specifications of the server, which is then sent to the central database and recorded in the table `host_info`.
+
+`host_usage.sh` is a script that is scheduled to run once a minute. It records the server's resource usage, specifically its CPU and memory utilization. This script also pushes this information to the database table `host_usage` for recording.
+
+## Usage 
+1. `psql_docker.sh`: there are two (2) input options for this script.
```
Comment (Jarvis-deprecated-account) [inline linux_sql/README.md:25]: fix markdown syntax

File: linux_sql/scripts/host_info.sh [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+# Save the CPU hardware into relevant field variables.
+cpu_number=$(echo "$lscpu_output" | egrep "^CPU\(s\):" | awk '{print $2}' | xargs)
+cpu_architecture=$(echo "$lscpu_output" | egrep "^Architecture:" | awk '{print $2}' | xargs)
+cpu_model=$(echo "$lscpu_output" | egrep "^Model name:" | awk -F':' '{print $2}' | xargs)
+cpu_mhz=$(echo "$lscpu_output" | egrep "^CPU MHz:" | awk -F':' '{print $2}' | xargs)
+l2_cache=$(echo "$lscpu_output" | egrep "^L2 cache:" | awk '{print $3}' | xargs)
+total_mem=$(echo "$meminfo" | egrep "MemTotal:" | awk '{print $2}' | xargs)
+timestamp=$(vmstat -t | awk '{if(NR==3) print $18" "$19}' | xargs)
+
```
Comment (Jarvis-deprecated-account) [inline linux_sql/scripts/host_info.sh:28]: you need an env var for psql psswd

File: linux_sql/scripts/host_info.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+# A script to populate host_info table.
+# Format: psql_host psql_port db_name psql_user psql_password
+
+# Setting CLI arguments to variables
+psql_host=$1
+psql_port=$2
+database=$3
+psql_user=$4
+psql_password=$5
+
```
Comment (Jarvis-deprecated-account) [inline linux_sql/scripts/host_info.sh:12]: check # of args

Comment (Jarvis-deprecated-account) [reply linux_sql/scripts/host_info.sh:12]: shell coding style

File: linux_sql/scripts/host_usage.sh [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+timestamp=$(vmstat -t | awk '{if(NR==3) print $18" "$19}' | xargs)
+
+# CPU utilization data.
+memory_free=$(echo "$vmstat" | awk '{if(NR==3) print $4}' | xargs)
+cpu_idle=$(echo "$disk" | awk '{if(NR==2) print 100-$5}' | xargs)
+cpu_kernel=$(echo "$meminfo" | egrep "^KernelStack:" | awk '{print $2}' | xargs)
+disk_io=$(vmstat --unit M | awk '{if(NR==3) print $9}' | xargs)
+disk_available=$(echo "$disk" | awk '{if(NR==3) print $4}' | xargs)
+
+export PGPASSWORD='$psql_password'; psql -h $hostname -U $psql_user -w $database_name -p $psql_port -c \
```
Comment (Jarvis-deprecated-account) [inline linux_sql/scripts/host_usage.sh:30]: ;
||
&&
|

File: linux_sql/scripts/psql_docker.sh [hunk 6]
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
+# PSQL_Docker.sh script
+# The purpose of this script is to start/stop Docker and start/stop a PostgreSQL container. 
+
```
Comment (Jarvis-deprecated-account) [inline linux_sql/scripts/psql_docker.sh:4]: assign args to vars

File: linux_sql/scripts/psql_docker.sh [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+if [ "$1" != "start" ] && [ "$1" != "stop" ];
+then
+    echo "Invalid command. Please run the script with either the command start followed by the database password, or stop."
+    exit 1
+fi
+
+# If the command 'stop' is given, stop the running container.
+if [ "$1" = "stop" ];
+then
+   docker container stop jrvs-psql
```
Comment (Jarvis-deprecated-account) [inline linux_sql/scripts/psql_docker.sh:15]: check  daemon

File: linux_sql/scripts/psql_docker.sh [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+if [ "$(docker container ls -a -f name=jrvs-psql | wc -l)" != "2" ];
+then
+    docker run --name jrvs-psql -e POSTGRES_PASSWORD=$PGPASSWORD -d -v pgdata:/var/lib/postgresql/data -p 5432:5432 postgres
+else
+    docker container start jrvs-psql
+echo "Docker container started!"
+fi
+
+sudo yum install -y postgresql
```
Comment (Jarvis-deprecated-account) [inline linux_sql/scripts/psql_docker.sh:66]: remove 66-71

File: linux_sql/sql/queries.sql [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+	) as avg_used_memory_percentage
+FROM 
+	host_usage AS usage
+	JOIN
+		host_info AS info
+		ON usage.host_id = info.id
+GROUP BY
+	usage.host_id
+	info.hostname,
+	info.total_mem,
```
Comment (Jarvis-deprecated-account) [inline linux_sql/sql/queries.sql:42]: do you need group by total_mem

**Review by ghost — CHANGES_REQUESTED** [review]

**Pull request closed without merging**